### ISSUE TITLE
Good disaggregation results?

### ISSUE THREAD

**Issue by nekovolta** [nilmtk/nilmtk#788]
Following issue #775, I am following the example, with all the appliances and the range of time for training and test in issue #775. In that example, I am using all the appliances for training and test. My question is that the results I got after the disaggregation seem not very good.

The sample rate that I used is: _sample rate = 1_ as it would be to use the maximum information available as the data granularity of the dataset is 1 Hz. Using higher _sample rates_ would lead to even worse results.

- **Hair dryer**

The results are horrible. There is a detected use of the _hair dryer_ when it is not used at all. On the other hand, the prediction says there is a lot of use of the _hair dryer_ when in reality it is not used that much. I would not consider a useful disaggregation feature this one.

![image](https://user-images.githubusercontent.com/29656053/59188045-0bfcd980-8b77-11e9-9d63-80d8aaa47286.png)

- **Kettle**

The results are similar to the case of the _hair dryer_

![image](https://user-images.githubusercontent.com/29656053/59188147-4bc3c100-8b77-11e9-8a44-065e534c544d.png)
 
- **Computer**

I think these are the worst results. I would not consider the prediction close to the real consumption/use of the computer based on ground truth data.

![image](https://user-images.githubusercontent.com/29656053/59188361-bffe6480-8b77-11e9-9992-a7f81760cb80.png)

- **Washing machine**

Better results than the cases commented previously of the _hair dryer_ or the _kettle_, but still very different to the ground truth data.

![image](https://user-images.githubusercontent.com/29656053/59188515-1c618400-8b78-11e9-9554-03b72fed305a.png)

- **Fridge**

These could be the better results, but as you can see in the picture, it is also far from close disaggregation.

![image](https://user-images.githubusercontent.com/29656053/59188670-7104ff00-8b78-11e9-8b13-311266bb2da5.png)

- **Freezer**

Comparing with the fridge, I would not consider these results as good:

![image](https://user-images.githubusercontent.com/29656053/59188879-00121700-8b79-11e9-85c8-0b8cb20f87fb.png)

I do not understand, as the freezer could have a behaviour similar to the fridge and the results are very different.

- **Coffe Maker**

It shows this error. Why?

**In conclusion**: What do you think of these results? Considering that the _sample rate_ is the maximum available: are these the better disaggregation results that the algorithm can offer?

**Comment by rajatgupta1234** [comment]
@nekovolta 
Can you please give details of code you used?
Like Dataset, Train and Test Window size, which algorithm you used etc. Please give a screenshot of your code.

**Comment by nekovolta** [comment]
@rajatgupta1234 

here you have the complete code I used, with all the details, even the "plot resolution" to get the same results.


```python
from __future__ import print_function, division
import time
from matplotlib import rcParams
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from six import iteritems
import warnings

warnings.filterwarnings('ignore')

rcParams['figure.figsize'] = (13, 6)

from nilmtk import DataSet, TimeFrame, MeterGroup, HDFDataStore
from nilmtk.disaggregate import CombinatorialOptimisation, FHMM
from nilmtk.utils import print_dict

# selection of the data for ECO dataset

eco = DataSet('/ECO/eco.h5')
elec = eco.buildings[1].elec
elec

# Dividing data into train and test set
train = DataSet('/ECO/eco.h5')
test = DataSet('/ECO/eco.h5')

# selection of building 1
building = 1

# format 'yyyy-mm-dd'
train.set_window(end="2012-07-01")
test.set_window(start="2012-07-01", end = "2012-07-15")

# MeterGroups for the train and test selection
train_elec = train.buildings[1].elec
test_elec = test.buildings[1].elec

# TRAINING AND DISAGGREGATION

# A function to disaggregate the mains data to constituent appliances and return the predictions

def predict(clf, test_elec, sample_period, timezone):
    pred = {}
    gt= {}
    
    # "ac_type" varies according to the dataset used. 
    # Make sure to use the correct ac_type before using the default parameters in this code.    
    for i, chunk in enumerate(test_elec.mains().load(physical_quantity = 'power', ac_type = 'active', sample_period=sample_period)):
        chunk_drop_na = chunk.dropna()
        pred[i] = clf.disaggregate_chunk(chunk_drop_na)
        gt[i]={}

        for meter in test_elec.submeters().meters:
            # Only use the meters that we trained on (this saves time!)    
            gt[i][meter] = next(meter.load(physical_quantity = 'power', ac_type = 'active', sample_period=sample_period))
        gt[i] = pd.DataFrame({k:v.squeeze() for k,v in iteritems(gt[i]) if len(v)}, index=next(iter(gt[i].values())).index).dropna()

    # If everything can fit in memory
    gt_overall = pd.concat(gt)
    gt_overall.index = gt_overall.index.droplevel()
    pred_overall = pd.concat(pred)
    pred_overall.index = pred_overall.index.droplevel()

    # Having the same order of columns
    gt_overall = gt_overall[pred_overall.columns]
    
    #Intersection of index
    gt_index_utc = gt_overall.index.tz_convert("UTC")
    pred_index_utc = pred_overall.index.tz_convert("UTC")
    common_index_utc = gt_index_utc.intersection(pred_index_utc)
    
    common_index_local = common_index_utc.tz_convert(timezone)
    gt_overall = gt_overall.ix[common_index_local]
    pred_overall = pred_overall.ix[common_index_local]
    appliance_labels = [m for m in gt_overall.columns.values]
    gt_overall.columns = appliance_labels
    pred_overall.columns = appliance_labels
    return gt_overall, pred_overall

# Train using 2 benchmarking algorithms - Combinatorial Optimisation (CO) and Factorial Hidden Markov Model (FHMM)
classifiers = {'FHMM':FHMM()}
predictions = {}
sample_period = 1
for clf_name, clf in classifiers.items():
    print("*"*20)
    print(clf_name)
    print("*" *20)
    start = time.time()
    clf.train(train_elec, sample_period=sample_period)
    end = time.time()
    print("Runtime =", end-start, "seconds.")
    gt, predictions[clf_name] = predict(clf, test_elec, sample_period, train.metadata['timezone'])

appliance_labels = [m.label() for m in gt.columns.values]

gt.columns = appliance_labels
predictions['FHMM'].columns = appliance_labels

# Taking a look at the ground truth of top 5 appliance power consumption
gt.head()
predictions['FHMM'].head()


predictions['FHMM']['Fridge'].head(300*60*4).plot(label="Pred")
gt['Fridge'].head(300*60*4).plot(label="GT")
plt.legend()
#

predictions['FHMM']['Freezer'].head(300*60*4).plot(label="Pred")
gt['Freezer'].head(300*60*4).plot(label="GT")
plt.legend()
#

predictions['FHMM']['Washing machine'].head(300*6000).plot(label="Pred")
gt['Washing machine'].head(300*6000).plot(label="GT")
plt.legend()
#

predictions['FHMM']['Computer'].head(300*6000).plot(label="Pred")
gt['Computer'].head(300*6000).plot(label="GT")
plt.legend()
#

predictions['FHMM']['Kettle'].head(300*6000).plot(label="Pred")
gt['Kettle'].head(300*6000).plot(label="GT")
plt.legend()
#

predictions['FHMM']['Hair dryer'].head(300*6000).plot(label="Pred")
gt['Hair dryer'].head(300*6000).plot(label="GT")
plt.legend()
```

**Comment by rajatgupta1234** [comment]
@nekovolta  
You used sample_period=1 and trained the algorithms on 1 month data. Let's see with sample_period=120 and 2 months data for training. This is better prediction. Same you can try with other devices. I will shortly come up with the Coffee Maker error issue. 
Fridge Prediction using CO.
![CO](https://user-images.githubusercontent.com/46721220/59770025-e41f1b80-92c4-11e9-8d27-bab5dde19e94.PNG)
Fridge Prediction using FHMM.
![FHMM](https://user-images.githubusercontent.com/46721220/59770092-05800780-92c5-11e9-93aa-094b534d7226.PNG)

**Comment by nekovolta** [comment]
@rajatgupta1234 

Do you consider that those results are fine? For instance, in the prediction using FHMM almost all the predictions find that the fridge is ON when the GT data says it was OFF, and also just the other way around: the prediction says it was OFF when it was ON

**Comment by rajatgupta1234** [comment]
@nekovolta , This is the code which I used. I didn't mention it earlier.
`from __future__ import print_function, division
import time
from matplotlib import rcParams
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from six import iteritems
import warnings

warnings.filterwarnings('ignore')
%matplotlib inline

rcParams['figure.figsize'] = (13,6)
import nilmtk
from nilmtk import DataSet, TimeFrame, MeterGroup, HDFDataStore
from nilmtk.disaggregate import CombinatorialOptimisation, FHMM
from nilmtk.utils import print_dict

train = DataSet('C:\\Users\\Rajat\\Desktop\\data\\eco.h5')
test = DataSet('C:\\Users\\Rajat\\Desktop\\data\\eco.h5')

building = 1

train.set_window(end="2012-08-01")
test.set_window(start="2012-08-01", end = "2012-08-15")

train_elec = train.buildings[1].elec
test_elec = test.buildings[1].elec

def predict(clf, test_elec, sample_period, timezone):
    pred = {}
    gt= {}
    
    # "ac_type" varies according to the dataset used. 
    # Make sure to use the correct ac_type before using the default parameters in this code.    
    for i, chunk in enumerate(test_elec.mains().load(physical_quantity = 'power', ac_type = 'active', sample_period=sample_period)):
        chunk_drop_na = chunk.dropna()
        pred[i] = clf.disaggregate_chunk(chunk_drop_na)
        gt[i]={}

        for meter in test_elec.submeters().meters:
            # Only use the meters that we trained on (this saves time!)    
            gt[i][meter] = next(meter.load(physical_quantity = 'power', ac_type = 'active', sample_period=sample_period))
        gt[i] = pd.DataFrame({k:v.squeeze() for k,v in iteritems(gt[i]) if len(v)}, index=next(iter(gt[i].values())).index).dropna()
        
    # If everything can fit in memory
    gt_overall = pd.concat(gt)
    gt_overall.index = gt_overall.index.droplevel()
    pred_overall = pd.concat(pred)
    pred_overall.index = pred_overall.index.droplevel()

    # Having the same order of columns
    gt_overall = gt_overall[pred_overall.columns]
    
    #Intersection of index
    gt_index_utc = gt_overall.index.tz_convert("UTC")
    pred_index_utc = pred_overall.index.tz_convert("UTC")
    common_index_utc = gt_index_utc.intersection(pred_index_utc)
    
    common_index_local = common_index_utc.tz_convert(timezone)
    gt_overall = gt_overall.ix[common_index_local]
    pred_overall = pred_overall.ix[common_index_local]
    appliance_labels = [m for m in gt_overall.columns.values]
    gt_overall.columns = appliance_labels
    pred_overall.columns = appliance_labels
    return gt_overall, pred_overall

classifiers = {'CO':CombinatorialOptimisation(), 'FHMM':FHMM()}
predictions = {}
sample_period = 120
for clf_name, clf in classifiers.items():
    print("*"*20)
    print(clf_name)
    print("*" *20)
    start = time.time()
    # Note that we have given the sample period to downsample the data to 1 minute. 
    # If instead of top_5 we wanted to train on all appliance, we would write 
    # fhmm.train(train_elec, sample_period=60)
    clf.train(train_elec, sample_period=sample_period)
    end = time.time()
    print("Runtime =", end-start, "seconds.")
    gt, predictions[clf_name] = predict(clf, test_elec, sample_period, train.metadata['timezone'])
   
appliance_labels = [m.label() for m in gt.columns.values]
gt.columns = appliance_labels
predictions['CO'].columns = appliance_labels
predictions['FHMM'].columns = appliance_labels

predictions['CO']['Fridge'].head(300).plot(label="Pred")
gt['Fridge'].head(300).plot(label="GT")
plt.legend()

predictions['FHMM']['Fridge'].head(300).plot(label="Pred")
gt['Fridge'].head(300).plot(label="GT")
plt.legend()
`

**Comment by rajatgupta1234** [comment]
@nekovolta 
Please don't judge the results by seeing just a window. See the Root Mean Square Values for comparison.
An Example Code and Output:

```python
from sklearn.metrics import mean_squared_error

def compute_rmse(gt, pred):
    from sklearn.metrics import mean_squared_error
    rms_error = {}
    for appliance in gt.columns:
        rms_error[appliance] = np.sqrt(mean_squared_error(gt[appliance], pred[appliance]))
    return pd.Series(rms_error)

def compute_rmse(gt, pred):
    from sklearn.metrics import mean_squared_error
    rms_error = {}
    for appliance in gt.columns:
        rms_error[appliance] = np.sqrt(mean_squared_error(gt[appliance], pred[appliance]))
    return pd.Series(rms_error) 
```

The Error for sample period = 1

![rmse_1](https://user-images.githubusercontent.com/46721220/59787457-16da0b80-92e7-11e9-8f71-8ebf44a3a9d2.PNG)

The Error for sample period = 60

![rmse_60](https://user-images.githubusercontent.com/46721220/59787478-278a8180-92e7-11e9-81cf-7f118b3f0771.PNG)

The Error for sample period = 120

![rmse_120](https://user-images.githubusercontent.com/46721220/59787495-3113e980-92e7-11e9-8b86-53d05155e8a2.PNG)

**Comment by nekovolta** [comment]
@rajatgupta1234 

Thank you for your comment. This is what I do not understand. Checking the RMSE values, in my opinion, it says something very similar to what I discuss graphically. Let me explain:

In the next figure, I remarked in red which case you get the worst result (in orange where you get the best result for those in red), in green in which case you get the best result and in blue when the results for all the cases are similar.

![Imagen2](https://user-images.githubusercontent.com/29656053/59910027-51d85e00-9411-11e9-847d-ad0d83281506.png)

As you can see, for **sample_period = 1**, just one case where the results are better than in the other cases, you get five instances where the results are similar (not significant improvement) and four cases where the results are even worse than in other cases. 

I understood that better resolution improves disaggregation results. I read several papers remarking the benefits of high-frequency resolution (kHz) to improve disaggregation features, although the high-frequency resolution has technical limitations and could be not practical in real cases, theoretically it leads to better results. But these results tell us just the opposite: if you improve the resolution the predictions are works or the same, **in comparison with a resolution of 120, which means a data/registration/power measurement each 2 minutes!!!** against a resolution of 1, with means a **a data/registration/power measurement each second!**
